LINUX/if_vmxnet3_netmap.h

```diff
@@ -0,0 +1,593 @@
+
+	if (err) {
+		vmxnet3_force_close(adapter);
+	}
+
+	return 0;
+}
+
+static int vmxnet3_netmap_unmap_pkt(u32 eop_idx, struct vmxnet3_tx_queue *tq,
+				    struct pci_dev *pdev)
+{
+	int entries = 0;
+
+	//
+	// no out of order completion
+	//
+
+	BUG_ON(tq->buf_info[eop_idx].sop_idx != tq->tx_ring.next2comp);
+	BUG_ON(VMXNET3_TXDESC_GET_EOP(&(tq->tx_ring.base[eop_idx].txd)) != 1);
+
+	BUG_ON(tq->buf_info[eop_idx].skb != NULL);
+
+	VMXNET3_INC_RING_IDX_ONLY(eop_idx, tq->tx_ring.size);
+
+	while (tq->tx_ring.next2comp != eop_idx) {
+		vmxnet3_unmap_tx_buf(tq->buf_info + tq->tx_ring.next2comp,
+				     pdev);
+
+		//
+		// update next2comp w/o tx_lock. Since we are marking more,
+		// instead of less, tx ring entries avail, the worst case is
+		// that the tx routine incorrectly re-queues a pkt due to
+		// insufficient tx ring entries.
+		//
+
+		vmxnet3_cmd_ring_adv_next2comp(&tq->tx_ring);
+		entries++;
+	}
+
+	return entries;
+}
+
+static int vmxnet3_netmap_tq_tx_complete(struct vmxnet3_tx_queue *tq,
+					 struct pci_dev *pdev)
+{
+	int completed = 0;
+	union Vmxnet3_GenericDesc *gdesc;
+
+	gdesc = tq->comp_ring.base + tq->comp_ring.next2proc;
+
+	while (VMXNET3_TCD_GET_GEN(&gdesc->tcd) == tq->comp_ring.gen) {
+		completed += vmxnet3_netmap_unmap_pkt(
+			VMXNET3_TCD_GET_TXIDX(&gdesc->tcd), tq, pdev);
+
+		vmxnet3_comp_ring_adv_next2proc(&tq->comp_ring);
+		gdesc = tq->comp_ring.base + tq->comp_ring.next2proc;
+	}
+
+	return completed;
+}
+
+static int vmxnet3_netmap_txsync(struct netmap_kring *kring, int flags)
+{
+#define kUseTwoTxDescForPacket 0
+#define kMinFreeTxDescForPacket (kUseTwoTxDescForPacket ? 2 : 1)
+
+	struct netmap_adapter *na = kring->na;
+	struct ifnet *ifp = na->ifp;
+	struct netmap_ring *ring = kring->ring;
+
+	u_int n;
+	u_int nm_i; // index into the netmap ring
+	int completed;
+	u_int deferred = 0;
+	u_int ring_nr = kring->ring_id;
+
+	u_int const lim = kring->nkr_num_slots - 1;
+	u_int const head = kring->rhead;
+
+	union Vmxnet3_GenericDesc *gdesc;
+	struct SOFTC_T *adapter = netdev_priv(ifp);
+	struct vmxnet3_tx_queue *tq = &adapter->tx_queue[ring_nr];
+
+	if (!netif_carrier_ok(ifp))
+		return 0;
+
+	//
+	// Free up the comp_descriptors aggressively
+	//
+
+	completed = vmxnet3_netmap_tq_tx_complete(tq, adapter->pdev);
+
+	//
+	// Reclaim buffers for completed transmissions
+	//
+
+	kring->nr_hwtail =
+		nm_prev(tq->comp_ring.next2proc, tq->comp_ring.size - 1);
+
+	//
+	// Process new packets to send
+	//
+
+	nm_i = kring->nr_hwcur;
+
+	if (nm_i != head) {
+		for (n = 0; nm_i != head; n++) {
+			u32 copy_size = 0;
+			int free_cmd_desc_count;
+			unsigned long lock_flags;
+
+			union Vmxnet3_GenericDesc *sop_txd;
+			union Vmxnet3_GenericDesc *eop_txd;
+
+			struct netmap_slot *slot = &ring->slot[nm_i];
+			u_int packet_len = slot->len;
+			void *packet_addr = NMB(na, slot);
+
+			slot->flags &= ~(NS_REPORT | NS_BUF_CHANGED);
+
+			spin_lock_irqsave(&tq->tx_lock, lock_flags);
```

Comment:
Is there anything else that is accessing this queue while it is in netmap mode?

Reply:
"vmxnet3_tq_tx_complete" may access this queue but I do not know if it can be called in netmap mode